File: langs/csharp/src/VMx/Messages/CollectionChangedMessage.cs
using System.Collections.Specialized;

namespace VMx.Messages;

/// <summary>
/// Default <see cref="ICollectionChangedMessage{T}"/> implementation.
/// Published by <see cref="VMx.Collections.ServicedObservableCollection{T}"/> to the hub.
/// See spec/21-collections.md §2 and ADR-0024.
/// </summary>
/// <typeparam name="T">Element type of the collection.</typeparam>
public sealed record CollectionChangedMessage<T>(
    object Sender,
    NotifyCollectionChangedAction Action,
    IReadOnlyList<T> NewItems,
    IReadOnlyList<T> OldItems,
    int Index) : ICollectionChangedMessage<T>
{
    /// <summary>Source position, or -1 when the action has no source.</summary>
    public int OldIndex { get; init; } = -1;

    /// <summary>Destination position, or -1 when the action has no destination.</summary>
    public int NewIndex { get; init; } = -1;

    /// <inheritdoc/>
    public object SenderObject => Sender;

    /// <inheritdoc/>
    /// <remarks>Derived from the sender's runtime type name; no separate name field per spec §2.4.</remarks>
    public string SenderName => Sender.GetType().Name;

    /// <summary>Factory for Add.</summary>
    public static CollectionChangedMessage<T> ForAdd(object sender, T item, int index)
        => new(sender, NotifyCollectionChangedAction.Add,
               new[] { item }, Array.Empty<T>(), index)
        {
            NewIndex = index,
        };

    /// <summary>Factory for Remove.</summary>
    public static CollectionChangedMessage<T> ForRemove(object sender, T item, int index)
        => new(sender, NotifyCollectionChangedAction.Remove,
               Array.Empty<T>(), new[] { item }, index)
        {
            OldIndex = index,
        };

    /// <summary>Factory for Replace.</summary>
    public static CollectionChangedMessage<T> ForReplace(object sender, T newItem, T oldItem, int index)
        => new(sender, NotifyCollectionChangedAction.Replace,
               new[] { newItem }, new[] { oldItem }, index)
        {
            OldIndex = index,
            NewIndex = index,
        };

    /// <summary>Factory for Move.</summary>
    public static CollectionChangedMessage<T> ForMove(object sender, T item, int oldIndex, int newIndex)
        => new(sender, NotifyCollectionChangedAction.Move,
               new[] { item }, new[] { item }, newIndex)
        {
            OldIndex = oldIndex,
            NewIndex = newIndex,
        };

    /// <summary>Factory for Reset.</summary>
    public static CollectionChangedMessage<T> ForReset(object sender)
        => new(sender, NotifyCollectionChangedAction.Reset,
               Array.Empty<T>(), Array.Empty<T>(), -1);
}

File: langs/swift/Sources/VMx/Collections/CollectionChangedMessage.swift
//
// CollectionChangedMessage — hub envelope emitted by ServicedObservableCollection.
//
// See spec/21-collections.md §2 and ADR-0024.
//
import Foundation

/// Hub message published by `ServicedObservableCollection<T>` on each mutation.
///
/// Conforms to `Message` so it is routable through `MessageHubProtocol`.
/// Subscribers may downcast `any Message` to `CollectionChangedMessage<T>` or
/// filter by identity (`sender === collection`).
public struct CollectionChangedMessage<T>: Message {
    /// The collection that emitted this message.
    public let senderObject: AnyObject
    /// Type name of the sender (e.g. `"ServicedObservableCollection"`).
    public let senderName: String
    /// The kind of mutation.
    public let action: CollectionChangedAction
    /// Items added or replacing an existing item (empty for remove/reset).
    public let newItems: [T]
    /// Items removed or being replaced (empty for add/reset).
    public let oldItems: [T]
    /// Insertion/removal/replacement index, or -1 when not applicable (reset).
    public let index: Int
    /// Position before the mutation, or -1 when there is no old position.
    public let oldIndex: Int
    /// Position after the mutation, or -1 when there is no new position.
    public let newIndex: Int

    /// Creates a collection-change message. The explicit old/new positions are
    /// defaulted so existing callers that provide only the legacy `index`
    /// remain source-compatible.
    public init(
        senderObject: AnyObject,
        senderName: String,
        action: CollectionChangedAction,
        newItems: [T],
        oldItems: [T],
        index: Int,
        oldIndex: Int = -1,
        newIndex: Int = -1
    ) {
        self.senderObject = senderObject
        self.senderName = senderName
        self.action = action
        self.newItems = newItems
        self.oldItems = oldItems
        self.index = index
        self.oldIndex = oldIndex
        self.newIndex = newIndex
    }

    /// Spec alias of `senderObject`. Matches the `sender` field name in
    /// C# / Python / TypeScript flavors.
    public var sender: AnyObject { senderObject }
}

// MARK: — Convenience factories

extension CollectionChangedMessage {
    /// An Add message for a single item appended at `index`.
    public static func forAdd(
        sender: AnyObject,
        senderName: String,
        item: T,
        index: Int
    ) -> CollectionChangedMessage<T> {
        .init(senderObject: sender, senderName: senderName,
              action: .add, newItems: [item], oldItems: [], index: index,
              oldIndex: -1, newIndex: index)
    }

    /// A Remove message for a single item that was at `index` before removal.
    public static func forRemove(
        sender: AnyObject,
        senderName: String,
        item: T,
        index: Int
    ) -> CollectionChangedMessage<T> {
        .init(senderObject: sender, senderName: senderName,
              action: .remove, newItems: [], oldItems: [item], index: index,
              oldIndex: index, newIndex: -1)
    }

    /// A Replace message for an item replaced in-place at `index`.
    public static func forReplace(
        sender: AnyObject,
        senderName: String,
        newItem: T,
        oldItem: T,
        index: Int
    ) -> CollectionChangedMessage<T> {
        .init(senderObject: sender, senderName: senderName,
              action: .replace, newItems: [newItem], oldItems: [oldItem], index: index,
              oldIndex: index, newIndex: index)
    }

    /// A Move message for an existing item relocated without replacement.
    public static func forMove(
        sender: AnyObject,
        senderName: String,
        item: T,
        from oldIndex: Int,
        to newIndex: Int
    ) -> CollectionChangedMessage<T> {
        .init(senderObject: sender, senderName: senderName,
              action: .move, newItems: [item], oldItems: [item], index: newIndex,
              oldIndex: oldIndex, newIndex: newIndex)
    }

    /// A Reset message (clear or batch update). `index` is -1.
    public static func forReset(
        sender: AnyObject,
        senderName: String
    ) -> CollectionChangedMessage<T> {
        .init(senderObject: sender, senderName: senderName,
              action: .reset, newItems: [], oldItems: [], index: -1,
              oldIndex: -1, newIndex: -1)
    }
}
